File: Core/fact_checker.py
"""
Fact Checker - FEVER-inspired fact verification system

Based on:
- FEVER: Fact Extraction and VERification (Thorne et al., 2018)
- Self-RAG: Learning to Retrieve, Generate, and Critique (Asai et al., 2023)
- Modern hallucination detection techniques

Verifies claims against retrieved evidence with confidence scoring.
"""

import logging
import re
from typing import List, Dict, Any, Optional
from dataclasses import dataclass, field
from enum import Enum

logger = logging.getLogger(__name__)


class VerificationLabel(str, Enum):
    """FEVER-style verification labels"""

    SUPPORTED = "SUPPORTED"  # Evidence directly supports claim
    REFUTED = "REFUTED"  # Evidence contradicts claim
    NOT_ENOUGH_INFO = "NOT_ENOUGH_INFO"  # No evidence either way
    PARTIALLY_SUPPORTED = "PARTIALLY_SUPPORTED"  # Some support, some contradiction
    UNVERIFIABLE = "UNVERIFIABLE"  # Claim can't be verified from evidence


class HallucinationLevel(str, Enum):
    """Likelihood that a claim is hallucinated (not in evidence)"""

    SUPPORTED = "supported"  # Clearly in evidence
    LIKELY_SUPPORTED = "likely_supported"  # Inferred from evidence
    UNCERTAIN = "uncertain"  # Could be inferred
    LIKELY_HALLUCINATED = "likely_hallucinated"  # Probably made up
    CLEARLY_HALLUCINATED = "clearly_hallucinated"  # Definitely made up


@dataclass
class Claim:
    """Single factual claim extracted from answer"""

    text: str
    subject: str  # Entity being described
    predicate: str  # What property/relation
    object_value: str  # The value/property
    sentence_idx: int


@dataclass
class Evidence:
    """Source evidence for evaluating a claim"""

    text: str
    source_name: str
    source_type: str
    source_date: Optional[str] = None
    relevance_score: float = 1.0


@dataclass
class VerificationResult:
    """Result of verifying a single claim"""

    claim: Claim
    label: VerificationLabel
    confidence: float  # 0-1
    supporting_evidence: List[Evidence] = field(default_factory=list)
    contradicting_evidence: List[Evidence] = field(default_factory=list)
    reasoning: str = ""
    hallucination_level: HallucinationLevel = HallucinationLevel.SUPPORTED


@dataclass
class AnswerVerification:
    """Complete verification for an answer"""

    answer: str
    claims: List[Claim]
    verifications: List[VerificationResult] = field(default_factory=list)
    overall_label: VerificationLabel = VerificationLabel.SUPPORTED
    confidence_score: float = 1.0  # Average confidence across claims
    hallucination_risk: float = 0.0  # 0-1 probability of hallucination
    key_supported_facts: List[str] = field(default_factory=list)
    key_unsupported_facts: List[str] = field(default_factory=list)
    critical_warnings: List[str] = field(default_factory=list)


class ClaimExtractor:
    """Extract factual claims from answer text using simple patterns"""

    # Sentence patterns that usually contain factual claims
    CLAIM_PATTERNS = [
        r"^(.+?)\s+(?:is|are|was|were)\s+(.+?)$",  # X is Y
        r"^(.+?)\s+(?:founded|created|born|died)\s+(?:in|on)?\s+(.+?)$",  # X founded in Y
        r"^(?:The|A)\s+(.+?)\s+(?:has|have|had)\s+(.+?)$",  # The X has Y
        r"^(.+?)\s+(?:said|stated|claims?|argues?)\s+(.+?)$",  # X said Y
    ]

    def extract_claims(self, answer: str) -> List[Claim]:
        """Extract individual factual claims from answer text"""
        claims = []
        sentences = re.split(r"[.!?]+", answer)

        for sent_idx, sentence in enumerate(sentences):
            sentence = sentence.strip()
            if not sentence or len(sentence) < 10:  # Skip short sentences
                continue

            # Try to match patterns
            for pattern in self.CLAIM_PATTERNS:
                match = re.match(pattern, sentence, re.IGNORECASE)
                if match:
                    groups = match.groups()
                    if len(groups) >= 2:
                        claim = Claim(
                            text=sentence,
                            subject=groups[0].strip(),
                            predicate=pattern.split("\\s+")[1],
                            object_value=groups[1].strip(),
                            sentence_idx=sent_idx,
                        )
                        claims.append(claim)
                        break

        return claims


class FactChecker:
    """FEVER-style fact verification against evidence"""

    def __init__(self):
        self.claim_extractor = ClaimExtractor()
        self.evidence_threshold = 0.3  # Minimum relevance to consider evidence

    def verify_answer(
        self,
        answer: str,
        evidence: List[Dict[str, Any]],
    ) -> AnswerVerification:
        """Verify all claims in answer against evidence"""

        # Extract claims
        claims = self.claim_extractor.extract_claims(answer)

        # Convert evidence to Evidence objects
        evidence_objects = [
            Evidence(
                text=e.get("text", ""),
                source_name=e.get("source", "Unknown"),
                source_type=e.get("type", "unknown"),
                source_date=e.get("date"),
                relevance_score=e.get("score", 0.5),
            )
            for e in evidence
        ]

        # Verify each claim
        verifications = []
        labels = []
        confidences = []
        hallucination_risks = []

        for claim in claims:
            verification = self._verify_claim(claim, evidence_objects)
            verifications.append(verification)
            labels.append(verification.label)
            confidences.append(verification.confidence)
            hallucination_risks.append(
                1.0
                if verification.hallucination_level
                in [
                    HallucinationLevel.CLEARLY_HALLUCINATED,
                    HallucinationLevel.LIKELY_HALLUCINATED,
                ]
                else 0.0
                if verification.hallucination_level == HallucinationLevel.SUPPORTED
                else 0.3
            )

        # Overall assessment
        overall_label = self._aggregate_labels(labels)
        avg_confidence = sum(confidences) / len(confidences) if confidences else 1.0
        avg_hallucination_risk = sum(hallucination_risks) / len(hallucination_risks) if hallucination_risks else 0.0

        # Identify key facts
        supported_facts = [
            v.claim.text
            for v in verifications
            if v.label in [VerificationLabel.SUPPORTED, VerificationLabel.PARTIALLY_SUPPORTED]
        ]
        unsupported_facts = [
            v.claim.text
            for v in verifications
            if v.label in [VerificationLabel.NOT_ENOUGH_INFO, VerificationLabel.UNVERIFIABLE]
        ]

        # Critical warnings
        critical_warnings = self._generate_warnings(verifications)

        return AnswerVerification(
            answer=answer,
            claims=claims,
            verifications=verifications,
            overall_label=overall_label,
            confidence_score=avg_confidence,
            hallucination_risk=avg_hallucination_risk,
            key_supported_facts=supported_facts[:5],
            key_unsupported_facts=unsupported_facts[:5],
            critical_warnings=critical_warnings,
        )

    def _verify_claim(
        self,
        claim: Claim,
        evidence: List[Evidence],
    ) -> VerificationResult:
        """Verify a single claim against evidence"""

        if not evidence:
            return VerificationResult(
                claim=claim,
                label=VerificationLabel.NOT_ENOUGH_INFO,
                confidence=0.0,
                hallucination_level=HallucinationLevel.CLEARLY_HALLUCINATED,
                reasoning="No evidence provided to verify claim",
            )

        supporting = []
        contradicting = []
        uncertain = []

        # Score claim against each piece of evidence
        for ev in evidence:
            score = self._similarity_score(claim, ev)

            if score > 0.7:  # Strong match
                supporting.append(ev)
            elif score > 0.4:  # Partial match
                uncertain.append(ev)
            elif self._is_contradictory(claim, ev):
                contradicting.append(ev)

        # Determine label
        if supporting:
            label = VerificationLabel.SUPPORTED
            hallucination_level = HallucinationLevel.SUPPORTED
            confidence = min(0.95, 0.5 + len(supporting) * 0.15)
        elif contradicting:
            label = VerificationLabel.REFUTED
            hallucination_level = HallucinationLevel.CLEARLY_HALLUCINATED
            confidence = 0.8
        elif uncertain:
            label = VerificationLabel.PARTIALLY_SUPPORTED
            hallucination_level = HallucinationLevel.LIKELY_SUPPORTED
            confidence = 0.5
        else:
            label = VerificationLabel.NOT_ENOUGH_INFO
            hallucination_level = HallucinationLevel.LIKELY_HALLUCINATED
            confidence = 0.2

        return VerificationResult(
            claim=claim,
            label=label,
            confidence=confidence,
            supporting_evidence=supporting,
            contradicting_evidence=contradicting,
            hallucination_level=hallucination_level,
            reasoning=self._generate_reasoning(claim, supporting, contradicting, uncertain),
        )

    def _similarity_score(self, claim: Claim, evidence: Evidence) -> float:
        """Score how much evidence supports a claim (0-1)"""
        # Simple token overlap + subject matching
        claim_tokens = set(claim.text.lower().split())
        evidence_tokens = set(evidence.text.lower().split())

        overlap = len(claim_tokens & evidence_tokens) / max(len(claim_tokens), 1)
        subject_match = 1.0 if claim.subject.lower() in evidence.text.lower() else 0.0

        # Weighted combination
        return (overlap * 0.5 + subject_match * 0.5) * evidence.relevance_score

    def _is_contradictory(self, claim: Claim, evidence: Evidence) -> bool:
        """Check if evidence contradicts claim"""
        # Look for explicit negations
        negation_words = ["not", "no", "never", "didn't", "doesn't", "won't"]

        claim.text.lower()
        evidence_lower = evidence.text.lower()

        # Very basic contradiction detection
        if claim.subject.lower() in evidence_lower:
            for neg in negation_words:
                if neg in evidence_lower and claim.predicate in evidence_lower:
                    return True

        return False

    def _aggregate_labels(self, labels: List[VerificationLabel]) -> VerificationLabel:
        """Aggregate multiple labels into overall label"""
        if not labels:
            return VerificationLabel.UNVERIFIABLE

        # Count label types
        from collections import Counter

        counts = Counter(labels)

        # REFUTED takes priority (direct contradiction)
        if counts.get(VerificationLabel.REFUTED, 0) > 0:
            return VerificationLabel.REFUTED

        # Then SUPPORTED
        if counts.get(VerificationLabel.SUPPORTED, 0) >= len(labels) * 0.6:
            return VerificationLabel.SUPPORTED

        # Then PARTIALLY_SUPPORTED
        if counts.get(VerificationLabel.PARTIALLY_SUPPORTED, 0) > 0:
            return VerificationLabel.PARTIALLY_SUPPORTED

        # Default to NOT_ENOUGH_INFO
        return VerificationLabel.NOT_ENOUGH_INFO

    def _generate_reasoning(
        self,
        claim: Claim,
        supporting: List[Evidence],
        contradicting: List[Evidence],
        uncertain: List[Evidence],
    ) -> str:
        """Generate human-readable reasoning"""
        if supporting:
            return f"Supported by {len(supporting)} source(s): {supporting[0].source_name}"
        elif contradicting:
            return f"Contradicted by {len(contradicting)} source(s): {contradicting[0].source_name}"
        elif uncertain:
            return f"Partially supported by {len(uncertain)} source(s)"
        else:
            return "No evidence found to verify this claim"

    def _generate_warnings(self, verifications: List[VerificationResult]) -> List[str]:
        """Generate critical warnings about answer quality"""
        warnings = []

        refuted = [v for v in verifications if v.label == VerificationLabel.REFUTED]
        unsupported = [v for v in verifications if v.label == VerificationLabel.NOT_ENOUGH_INFO]
        hallucinated = [
            v
            for v in verifications
            if v.hallucination_level
            in [
                HallucinationLevel.CLEARLY_HALLUCINATED,
                HallucinationLevel.LIKELY_HALLUCINATED,
            ]
        ]

        if refuted:
            warnings.append(f"⚠️ **CRITICAL**: {len(refuted)} claim(s) are directly contradicted by evidence!")

        if hallucinated:
            warnings.append(f"⚠️ **HIGH RISK**: {len(hallucinated)} claim(s) may be unsupported or hallucinated")

        if unsupported and len(unsupported) > len(verifications) * 0.3:
            warnings.append(f"⚠️ **Low Evidence**: {len(unsupported)} claim(s) have insufficient evidence")

        return warnings

File: rust_output/src/fact_checker.rs
/// Fact Checker - FEVER-inspired fact verification system
/// 
/// Based on:
/// - FEVER: Fact Extraction and VERification (Thorne et al., 2018)
/// - Self-RAG: Learning to Retrieve, Generate, and Critique (Asai et al., 2023)
/// - Modern hallucination detection techniques
/// 
/// Verifies claims against retrieved evidence with confidence scoring.

use anyhow::{Result, Context};
use regex::Regex;
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use std::collections::HashSet;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// FEVER-style verification labels
#[derive(Debug, Clone)]
pub struct VerificationLabel {
}

/// Likelihood that a claim is hallucinated (not in evidence)
#[derive(Debug, Clone)]
pub struct HallucinationLevel {
}

/// Single factual claim extracted from answer
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Claim {
    pub text: String,
    pub subject: String,
    pub predicate: String,
    pub object_value: String,
    pub sentence_idx: i64,
}

/// Source evidence for evaluating a claim
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Evidence {
    pub text: String,
    pub source_name: String,
    pub source_type: String,
    pub source_date: Option<String>,
    pub relevance_score: f64,
}

/// Result of verifying a single claim
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VerificationResult {
    pub claim: Claim,
    pub label: VerificationLabel,
    pub confidence: f64,
    pub supporting_evidence: Vec<Evidence>,
    pub contradicting_evidence: Vec<Evidence>,
    pub reasoning: String,
    pub hallucination_level: HallucinationLevel,
}

/// Complete verification for an answer
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnswerVerification {
    pub answer: String,
    pub claims: Vec<Claim>,
    pub verifications: Vec<VerificationResult>,
    pub overall_label: VerificationLabel,
    pub confidence_score: f64,
    pub hallucination_risk: f64,
    pub key_supported_facts: Vec<String>,
    pub key_unsupported_facts: Vec<String>,
    pub critical_warnings: Vec<String>,
}

/// Extract factual claims from answer text using simple patterns
#[derive(Debug, Clone)]
pub struct ClaimExtractor {
}

impl ClaimExtractor {
    /// Extract individual factual claims from answer text
    pub fn extract_claims(&mut self, answer: String) -> Vec<Claim> {
        // Extract individual factual claims from answer text
        let mut claims = vec![];
        let mut sentences = re::split("[.!?]+".to_string(), answer);
        for (sent_idx, sentence) in sentences.iter().enumerate().iter() {
            let mut sentence = sentence.trim().to_string();
            if (!sentence || sentence.len() < 10) {
                continue;
            }
            for pattern in self.CLAIM_PATTERNS.iter() {
                let mut r#match = regex::Regex::new(&pattern).unwrap().is_match(&sentence);
                if r#match {
                    let mut groups = r#match.groups();
                    if groups.len() >= 2 {
                        let mut claim = Claim(/* text= */ sentence, /* subject= */ groups[0].trim().to_string(), /* predicate= */ pattern.split("\\s+".to_string()).map(|s| s.to_string()).collect::<Vec<String>>()[1], /* object_value= */ groups[1].trim().to_string(), /* sentence_idx= */ sent_idx);
                        claims.push(claim);
                        break;
                    }
                }
            }
        }
        claims
    }
}

/// FEVER-style fact verification against evidence
#[derive(Debug, Clone)]
pub struct FactChecker {
    pub claim_extractor: ClaimExtractor,
    pub evidence_threshold: f64,
}

impl FactChecker {
    pub fn new() -> Self {
        Self {
            claim_extractor: ClaimExtractor(),
            evidence_threshold: 0.3_f64,
        }
    }
    /// Verify all claims in answer against evidence
    pub fn verify_answer(&mut self, answer: String, evidence: Vec<HashMap<String, Box<dyn std::any::Any>>>) -> AnswerVerification {
        // Verify all claims in answer against evidence
        let mut claims = self.claim_extractor.extract_claims(answer);
        let mut evidence_objects = evidence.iter().map(|e| Evidence(/* text= */ e.get(&"text".to_string()).cloned().unwrap_or("".to_string()), /* source_name= */ e.get(&"source".to_string()).cloned().unwrap_or("Unknown".to_string()), /* source_type= */ e.get(&"type".to_string()).cloned().unwrap_or("unknown".to_string()), /* source_date= */ e.get(&"date".to_string()).cloned(), /* relevance_score= */ e.get(&"score".to_string()).cloned().unwrap_or(0.5_f64))).collect::<Vec<_>>();
        let mut verifications = vec![];
        let mut labels = vec![];
        let mut confidences = vec![];
        let mut hallucination_risks = vec![];
        for claim in claims.iter() {
            let mut verification = self._verify_claim(claim, evidence_objects);
            verifications.push(verification);
            labels.push(verification.label);
            confidences.push(verification.confidence);
            hallucination_risks.push(if vec![HallucinationLevel.CLEARLY_HALLUCINATED, HallucinationLevel.LIKELY_HALLUCINATED].contains(&verification.hallucination_level) { 1.0_f64 } else { if verification.hallucination_level == HallucinationLevel.SUPPORTED { 0.0_f64 } else { 0.3_f64 } });
        }
        let mut overall_label = self._aggregate_labels(labels);
        let mut avg_confidence = if confidences { (confidences.iter().sum::<i64>() / confidences.len()) } else { 1.0_f64 };
        let mut avg_hallucination_risk = if hallucination_risks { (hallucination_risks.iter().sum::<i64>() / hallucination_risks.len()) } else { 0.0_f64 };
        let mut supported_facts = verifications.iter().filter(|v| vec![VerificationLabel.SUPPORTED, VerificationLabel.PARTIALLY_SUPPORTED].contains(&v.label)).map(|v| v.claim.text).collect::<Vec<_>>();
        let mut unsupported_facts = verifications.iter().filter(|v| vec![VerificationLabel.NOT_ENOUGH_INFO, VerificationLabel.UNVERIFIABLE].contains(&v.label)).map(|v| v.claim.text).collect::<Vec<_>>();
        let mut critical_warnings = self._generate_warnings(verifications);
        AnswerVerification(/* answer= */ answer, /* claims= */ claims, /* verifications= */ verifications, /* overall_label= */ overall_label, /* confidence_score= */ avg_confidence, /* hallucination_risk= */ avg_hallucination_risk, /* key_supported_facts= */ supported_facts[..5], /* key_unsupported_facts= */ unsupported_facts[..5], /* critical_warnings= */ critical_warnings)
    }
    /// Verify a single claim against evidence
    pub fn _verify_claim(&mut self, claim: Claim, evidence: Vec<Evidence>) -> VerificationResult {
        // Verify a single claim against evidence
        if !evidence {
            VerificationResult(/* claim= */ claim, /* label= */ VerificationLabel.NOT_ENOUGH_INFO, /* confidence= */ 0.0_f64, /* hallucination_level= */ HallucinationLevel.CLEARLY_HALLUCINATED, /* reasoning= */ "No evidence provided to verify claim".to_string())
        }
        let mut supporting = vec![];
        let mut contradicting = vec![];
        let mut uncertain = vec![];
        for ev in evidence.iter() {
            let mut score = self._similarity_score(claim, ev);
            if score > 0.7_f64 {
                supporting.push(ev);
            } else if score > 0.4_f64 {
                uncertain.push(ev);
            } else if self._is_contradictory(claim, ev) {
                contradicting.push(ev);
            }
        }
        if supporting {
            let mut label = VerificationLabel.SUPPORTED;
            let mut hallucination_level = HallucinationLevel.SUPPORTED;
            let mut confidence = 0.95_f64.min((0.5_f64 + (supporting.len() * 0.15_f64)));
        } else if contradicting {
            let mut label = VerificationLabel.REFUTED;
            let mut hallucination_level = HallucinationLevel.CLEARLY_HALLUCINATED;
            let mut confidence = 0.8_f64;
        } else if uncertain {
            let mut label = VerificationLabel.PARTIALLY_SUPPORTED;
            let mut hallucination_level = HallucinationLevel.LIKELY_SUPPORTED;
            let mut confidence = 0.5_f64;
        } else {
            let mut label = VerificationLabel.NOT_ENOUGH_INFO;
            let mut hallucination_level = HallucinationLevel.LIKELY_HALLUCINATED;
            let mut confidence = 0.2_f64;
        }
        VerificationResult(/* claim= */ claim, /* label= */ label, /* confidence= */ confidence, /* supporting_evidence= */ supporting, /* contradicting_evidence= */ contradicting, /* hallucination_level= */ hallucination_level, /* reasoning= */ self._generate_reasoning(claim, supporting, contradicting, uncertain))
    }
    /// Score how much evidence supports a claim (0-1)
    pub fn _similarity_score(&self, claim: Claim, evidence: Evidence) -> f64 {
        // Score how much evidence supports a claim (0-1)
        let mut claim_tokens = claim.text.to_lowercase().split_whitespace().map(|s| s.to_string()).collect::<Vec<String>>().into_iter().collect::<HashSet<_>>();
        let mut evidence_tokens = evidence.text.to_lowercase().split_whitespace().map(|s| s.to_string()).collect::<Vec<String>>().into_iter().collect::<HashSet<_>>();
        let mut overlap = ((claim_tokens & evidence_tokens).len() / claim_tokens.len().max(1));
        let mut subject_match = if evidence.text.to_lowercase().contains(&claim.subject.to_lowercase()) { 1.0_f64 } else { 0.0_f64 };
        (((overlap * 0.5_f64) + (subject_match * 0.5_f64)) * evidence.relevance_score)
    }
    /// Check if evidence contradicts claim
    pub fn _is_contradictory(&self, claim: Claim, evidence: Evidence) -> bool {
        // Check if evidence contradicts claim
        let mut negation_words = vec!["not".to_string(), "no".to_string(), "never".to_string(), "didn't".to_string(), "doesn't".to_string(), "won't".to_string()];
        claim.text.to_lowercase();
        let mut evidence_lower = evidence.text.to_lowercase();
        if evidence_lower.contains(&claim.subject.to_lowercase()) {
            for neg in negation_words.iter() {
                if (evidence_lower.contains(&neg) && evidence_lower.contains(&claim.predicate)) {
                    true
                }
            }
        }
        false
    }
    /// Aggregate multiple labels into overall label
    pub fn _aggregate_labels(&self, labels: Vec<VerificationLabel>) -> VerificationLabel {
        // Aggregate multiple labels into overall label
        if !labels {
            VerificationLabel.UNVERIFIABLE
        }
        // TODO: from collections import Counter
        let mut counts = Counter(labels);
        if counts.get(&VerificationLabel.REFUTED).cloned().unwrap_or(0) > 0 {
            VerificationLabel.REFUTED
        }
        if counts.get(&VerificationLabel.SUPPORTED).cloned().unwrap_or(0) >= (labels.len() * 0.6_f64) {
            VerificationLabel.SUPPORTED
        }
        if counts.get(&VerificationLabel.PARTIALLY_SUPPORTED).cloned().unwrap_or(0) > 0 {
            VerificationLabel.PARTIALLY_SUPPORTED
        }
        VerificationLabel.NOT_ENOUGH_INFO
    }
    /// Generate human-readable reasoning
    pub fn _generate_reasoning(&self, claim: Claim, supporting: Vec<Evidence>, contradicting: Vec<Evidence>, uncertain: Vec<Evidence>) -> String {
        // Generate human-readable reasoning
        if supporting {
            format!("Supported by {} source(s): {}", supporting.len(), supporting[0].source_name)
        } else if contradicting {
            format!("Contradicted by {} source(s): {}", contradicting.len(), contradicting[0].source_name)
        } else if uncertain {
            format!("Partially supported by {} source(s)", uncertain.len())
        } else {
            "No evidence found to verify this claim".to_string()
        }
    }
    /// Generate critical warnings about answer quality
    pub fn _generate_warnings(&self, verifications: Vec<VerificationResult>) -> Vec<String> {
        // Generate critical warnings about answer quality
        let mut warnings = vec![];
        let mut refuted = verifications.iter().filter(|v| v.label == VerificationLabel.REFUTED).map(|v| v).collect::<Vec<_>>();
        let mut unsupported = verifications.iter().filter(|v| v.label == VerificationLabel.NOT_ENOUGH_INFO).map(|v| v).collect::<Vec<_>>();
        let mut hallucinated = verifications.iter().filter(|v| vec![HallucinationLevel.CLEARLY_HALLUCINATED, HallucinationLevel.LIKELY_HALLUCINATED].contains(&v.hallucination_level)).map(|v| v).collect::<Vec<_>>();
        if refuted {
            warnings.push(format!("⚠️ **CRITICAL**: {} claim(s) are directly contradicted by evidence!", refuted.len()));
        }
        if hallucinated {
            warnings.push(format!("⚠️ **HIGH RISK**: {} claim(s) may be unsupported or hallucinated", hallucinated.len()));
        }
        if (unsupported && unsupported.len() > (verifications.len() * 0.3_f64)) {
            warnings.push(format!("⚠️ **Low Evidence**: {} claim(s) have insufficient evidence", unsupported.len()));
        }
        warnings
    }
}
